### DISCUSSION TITLE
Are there problems with declaring a state provider inside a widget

### DISCUSSION THREAD

**Discussion by titomitto** [rrousselGit/riverpod#2847] [Q&A] [answered]
@rrousselGit Since most of the examples I see a StateProvider being declared outside the widget, are there any problems with declaring it inside a widget? This is a screenshot of my code, check the `passwordVisibilityProvider`:

<img width="960" alt="image" src="https://github.com/rrousselGit/riverpod/assets/18606080/d49fc830-3548-4336-b761-91680cf27183">

This is only for cases where I just need to use it inside one widget only

**Comment by rrousselGit** [comment, marked as answer]
You're not respecting the rules of providers.
They should always be top-level variables.

Sounds like you don't want a provider. Maybe you're looking for flutter_hooks?

**Reply by titomitto** [reply]
Thanks. I'll explore flutter_hooks more on my personal projects, for this I'll just go the riverpod way as you suggested
